heron/schedulers/src/java/com/twitter/heron/scheduler/marathon/MarathonController.java

```diff
@@ -0,0 +1,198 @@
+    } catch (IOException ex) {
+      throw new RuntimeException("Failed to create connection for " + uri);
+    }
+
+    return connection;
+  }
+
+  public boolean killTopology() {
+    // Setup Connection
+    String topologyURI = String.format("%s/v2/groups/%s", this.marathonURI, this.topologyName);
+    HttpURLConnection conn = createHttpConnection(topologyURI);
+
+    try {
+      conn.setRequestMethod("DELETE");
+    } catch (ProtocolException e) {
+      LOG.log(Level.SEVERE, "Failed to set delete request: ", e);
+      conn.disconnect();
+      return false;
+    }
+
+    try {
+      conn.connect();
+    } catch (IOException ex) {
+      LOG.log(Level.SEVERE, "Failed to connect: ", ex);
+      return false;
+    } finally {
+      conn.disconnect();
+    }
+
+    // Check response
+    try {
+      int responseCode = conn.getResponseCode();
+      if (responseCode != HttpURLConnection.HTTP_OK) {
+        throw new RuntimeException("Failed to kill topology");
+      } else {
+        LOG.log(Level.INFO, "Topology killed successfully");
+      }
+    } catch (IOException ex) {
+      LOG.log(Level.SEVERE, "Failed to get response code");
+      return false;
+    } finally {
+      conn.disconnect();
+    }
+
+    return true;
+  }
+
+  public boolean restartApp(int appId) {
+    if (appId == -1) {
+      // TODO (nlu): implement restart all
+      throw new RuntimeException("Restart all containers not supported yet");
+    }
+
+    // Setup Connection
+    String restartRequest = String.format("%s/v2/apps/%s/%d/restart",
+        this.marathonURI, this.topologyName, appId);
+    HttpURLConnection conn = createHttpConnection(restartRequest);
+
+    try {
+      conn.setRequestMethod("POST");
+    } catch (ProtocolException e) {
+      LOG.log(Level.SEVERE, "Failed to set post request: ", e);
+      conn.disconnect();
+      return false;
+    }
+
+    try {
+      conn.connect();
+    } catch (IOException ex) {
+      LOG.log(Level.SEVERE, "Failed to connect: ", ex);
+      return false;
+    } finally {
+      conn.disconnect();
+    }
+
+    // Check response
+    try {
+      int responseCode = conn.getResponseCode();
+      if (responseCode != HttpURLConnection.HTTP_OK) {
+        String msg = String.format("Failed to restart container %d" + appId);
+        throw new RuntimeException(msg);
+      } else {
+        String msg = String.format("Container %s restarted successfully", appId);
+        LOG.log(Level.INFO, msg);
+      }
+    } catch (IOException ex) {
+      LOG.log(Level.SEVERE, "Failed to get response code");
+      return false;
+    } finally {
+      conn.disconnect();
+    }
+
+    return true;
+  }
+
+  // submit a topology as a group, containers as apps in the group
+  public boolean submitTopology(String appConf) {
+    if (this.isVerbose) {
+      LOG.log(Level.INFO, "Topology conf is: " + appConf);
+    }
+
+    // Setup Connection
+    String schedulerURI = String.format("%s/v2/groups", this.marathonURI);
```

Comment:
@nlu90 if you refactor the methods it would be good to keep the connection creation (preferably using a connection pool) within the class which is doing all the additional work of setting Request-type, etc. That might also mean not having the flexibility to use static methods.